### DISCUSSION TITLE
Add spinners to input components

### DISCUSSION THREAD

**Discussion by NemirStephan** [Legal-and-General/canopy#267] [Ideas]
There are instances where changing a value of an input might trigger an action such as saving to a back end database. It would be useful to have a 'busy' state added to all form inputs that would display an inline spinner.

Alternatively, all inputs could have a content area content like spinners can be projected in.

**Comment by andy-polhill** [comment]
Hi @NemirStephan thanks for raising this, do we have the designs at all for it? I was just wondering if the spinner is to go inside or outside of the input?

If it's inside there is a chance we can already use the 'add-on' button style with a spinner.
https://legal-and-general-canopy.netlify.app/?path=/story/components-form-input--with-button-suffix

**Reply by NemirStephan** [reply]
We don't have the designs just yet but can get them soon. For a text input, the 'add-on' style could work, but we also need this for selects, segments and in future, other inputs.

**Reply by alexcanning** [reply]
There may be other considerations related to this behaviour i.e. do we need to wait before displaying the spinner, or submitting data? Should the element be disabled while the spinner is visible? Latter point might go against our principles of not disabling components, but you get my drift.

**Reply by NemirStephan** [reply]
We can add t new input properties to the Select and Segment components (we can add these to other inputs as and when needed). I suggest the following API. I've chosen the word 'busy' as it is more generic than 'saving' or 'loading'

| Name                            | Description                   | Type       | Default | Required |
| ---------------------- | ---------------------- | -------- | --------| --------- |
| hasBusyState               | Indicates the component will be showing a spinner* | boolean | false | No |
| busy               | Shows the spinner when true  | boolean | false | No |

*if hasBusyState is true, then on mobile in full width, the input will leave some space to the right to allow for the spinner to be displayed without having to resize the input

##### Example of select component when `busy === true`
<img width="323" alt="Screenshot 2021-03-23 at 17 03 28" src="https://user-images.githubusercontent.com/22469863/112187145-bc778880-8bf9-11eb-974a-dad5956a7b82.png">

**Reply by elenagarrone** [reply]
Apologies but this sounds like bad user experience to me and dev wise you wouldn't want to keep pushing changes to the BE. You would want a user to confirm that they want to change something through a submit button first - I think that's what they would expect. I would feel weird to change something in a form and then not have the submit button, no?

Also form an accessibility point of view the submit button is a better choice.


It's different if the input is something like a search box or a filter that just fetches the correct data from a redux store or similar. In that case I would personally still prefer having the spinner where the updated content will appear.

**Reply by michelleallendavey** [reply]
<img width="157" alt="Screenshot 2021-03-26 at 11 47 56" src="https://user-images.githubusercontent.com/70210216/112627345-64bf6400-8e29-11eb-9a7e-41039494b467.png">
<img width="362" alt="Screenshot 2021-03-26 at 11 49 40" src="https://user-images.githubusercontent.com/70210216/112627346-6557fa80-8e29-11eb-98ff-09ed8a32f064.png">

@NemirStephan screenshots of the segment spinner as promised

@elenagarrone – definitely get where you're coming from! The autosave on things like segments came from some moderated testing we did where people wanted a save button where they'd inputted something themselves (text input for phone, email etc) but when changing something like post preferences, they tried to navigate away without saving because they assumed it automatically saved. Definitely worth an accessibility conversation though – I'm away next week but perhaps Seb can pick it up (looks like he's not on Github, will Slack him)

**Reply by michelleallendavey** [reply]
Hello again – picking this up. I'm definitely up for a discussion on rethinking the location of the spinner – just needs to work for segments, select dropdowns etc – anything that's not a text input. What were the accessibility concerns on not having the save button @elenagarrone? Maybe you and @NemirStephan can have a chat about how it would work – the messages that would come up and where, etc
